### PR TITLE
feat: add cloudConfig support in member-agent

### PR DESCRIPTION
**Pull request by britaniar**
### Description of your changes

Fixes #

I have: updated helm chart for member-agent to support cloud config

- [x] Run `make reviewable` to ensure this PR is ready for review.

### CODE REVIEW

File: pkg/utils/controller/controller.go [hunk 1]
```diff
@@ -171,7 +171,7 @@ func NewController(Name string, KeyFunc KeyFunc, ReconcileFunc ReconcileFunc, ra
 		name:          Name,
 		keyFunc:       KeyFunc,
 		reconcileFunc: ReconcileFunc,
-		queue:         workqueue.NewNamedRateLimitingQueue(rateLimiter, Name),
+		queue:         workqueue.NewRateLimitingQueue(rateLimiter),
```
Comment (zhiying-lin) [inline pkg/utils/controller/controller.go:174]: we should not remove the name here. Please use NewRateLimitingQueueWithConfig instead

File: test/e2e/setup.sh [hunk 2]
```diff
@@ -25,6 +25,7 @@ export HUB_AGENT_IMAGE="${HUB_AGENT_IMAGE:-hub-agent}"
 export MEMBER_AGENT_IMAGE="${MEMBER_AGENT_IMAGE:-member-agent}"
 export REFRESH_TOKEN_IMAGE="${REFRESH_TOKEN_IMAGE:-refresh-token}"
 export PROPERTY_PROVIDER="${PROPERTY_PROVIDER:-azure}"
+export CLOUD_CONFIG_FILE="${CLOUD_CONFIG_FILE:-"/etc/kubernetes/provider/config.json"}"
```
Comment (zhiying-lin) [inline test/e2e/setup.sh:28]: the default path should be the "azure_valid_config.yaml" under the test/e2e, right?

File: pkg/propertyprovider/azure/cloudconfig/config.go [hunk 3]
```diff
@@ -0,0 +1,155 @@
+package cloudconfig
```
Comment (zhiying-lin) [inline pkg/propertyprovider/azure/cloudconfig/config.go:1]: Wantong will create a package in this repo. please use that one.

File: pkg/propertyprovider/azure/provider.go [hunk 4]
```diff
@@ -69,6 +70,9 @@ type PropertyProvider struct {
 	// The controller manager in use by the Azure property provider; this field is mostly reserved for
 	// testing purposes.
 	mgr ctrl.Manager
+
+	// The cloud configuration in use by the Azure property provider.
+	cloudConfig cloudconfig.CloudConfig
```
Comment (zhiying-lin) [inline pkg/propertyprovider/azure/provider.go:75]: what you need is an azure client. It's better to initialize the client in the main func, so that it can be shared by the whole component.

here we don't need to pass in the whole cloudConfig

File: cmd/memberagent/main.go [hunk 5]
```diff
@@ -317,7 +320,7 @@ func Start(ctx context.Context, hubCfg, memberConfig *rest.Config, hubOpts, memb
 	discoverClient := discovery.NewDiscoveryClientForConfigOrDie(memberConfig)
 
 	if *enableV1Alpha1APIs {
-		gvk := workv1alpha1.SchemeGroupVersion.WithKind(workv1alpha1.AppliedWorkKind)
+		gvk := schema.GroupVersionKind{Group: workv1alpha1.GroupVersion.Group, Version: workv1alpha1.GroupVersion.Version, Kind: workv1alpha1.AppliedWorkKind}
```
[the remark this answers is not in the document]
Comment (britaniar) [reply cmd/memberagent/main.go:323]: Said it was deprecated
image

File: test/e2e/setup.sh (hunk 2 above)
Comment (britaniar) [reply test/e2e/setup.sh:28]: azure_valid_config.yaml will be used to upload all the values like it does in line 191. These values will be imported into the secret as a file located in /etc/kubernetes/provider/config.json

File: charts/member-agent/values.yaml [hunk 6]
```diff
@@ -35,6 +35,17 @@ config:
   identityKey: "identity-key-path"
   identityCert: "identity-cert-path"
   CABundle: "ca-bundle-path"
+  cloudConfig:
+    cloud: ""
+    tenantId: ""
+    subscriptionId: ""
+    useManagedIdentityExtension: false
+    userAgent: ""
```
Comment (jwtty) [inline charts/member-agent/values.yaml:43]: We also need userAssignedIdentityID, aadClientId, aadClientSecret, and resourceGroup 4 fields

File: pkg/propertyprovider/azure/cloudconfig/config.go (hunk 3 above)
Comment (jwtty) [reply pkg/propertyprovider/azure/cloudconfig/config.go:1]: Let's just merge this one.

File: pkg/propertyprovider/azure/cloudconfig/config.go [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"sigs.k8s.io/cloud-provider-azure/pkg/azclient"
+	"sigs.k8s.io/cloud-provider-azure/pkg/azclient/policy/ratelimit"
+	"sigs.k8s.io/cloud-provider-azure/pkg/consts"
+)
+
+const (
+	// DefaultUserAgent is the default user agent string to access Azure resources.
+	DefaultUserAgent = "fleet-member-agent"
```
Comment (jwtty) [inline pkg/propertyprovider/azure/cloudconfig/config.go:17]: Let's not keep a default userAgent here, since this module will be shared.

File: pkg/propertyprovider/azure/cloudconfig/config.go [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+)
+
+// CloudConfig holds the configuration parsed from the --cloud-config flag
+type RateLimitConfig ratelimit.Config
+type CloudConfig struct {
+	azclient.ARMClientConfig `json:",inline" mapstructure:",squash"`
+	azclient.AzureAuthConfig `json:",inline" mapstructure:",squash"`
+	*RateLimitConfig         `json:",inline" mapstructure:",squash"`
+	// name of cluster
```
Comment (jwtty) [inline pkg/propertyprovider/azure/cloudconfig/config.go:26]: Please also add a resourceGroup field.

File: pkg/propertyprovider/azure/cloudconfig/config.go [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+	contents, err := io.ReadAll(configReader)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read cloud cloudconfig file: %w, file path: %s", err, filePath)
+	}
+
+	if err := yaml.Unmarshal(contents, &config); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal cloud cloudconfig: %w, file path: %s", err, filePath)
+	}
+
+	if err := config.defaultAndValidate(); err != nil {
```
Comment (jwtty) [inline pkg/propertyprovider/azure/cloudconfig/config.go:64]: Could you merge my change in PR https://github.com/Azure/fleet-networking/pull/205?

File: pkg/propertyprovider/azure/cloudconfig/config.go [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+	// name of the virtual network of cluster
+	VnetName string `json:"vnetName,omitempty" mapstructure:"vnetName,omitempty"`
+	// name of the resource group where the virtual network is deployed
+	VnetResourceGroup string `json:"vnetResourceGroup,omitempty" mapstructure:"vnetResourceGroup,omitempty"`
+	// Enable exponential backoff to manage resource request retries
+	CloudProviderBackoff bool `json:"cloudProviderBackoff,omitempty" mapstructure:"cloudProviderBackoff,omitempty"`
+}
+
+// LoadCloudConfigFromFile loads the cloud config from the specified file
+func LoadCloudConfigFromFile(filePath string) (*CloudConfig, error) {
```
Comment (jwtty) [inline pkg/propertyprovider/azure/cloudconfig/config.go:43]: We can add another exported function `SetUserAgent` to allow setting userAgent for different components. I already updated the fleet-networking PR.

File: charts/member-agent/values.yaml [hunk 11]
```diff
@@ -35,6 +35,17 @@ config:
   identityKey: "identity-key-path"
   identityCert: "identity-cert-path"
   CABundle: "ca-bundle-path"
+  cloudConfig:
```
[the remark this answers is not in the document]
Comment (jwtty) [reply charts/member-agent/values.yaml:38]: Feel free to refer to the README in the fleet-networking PR.

File: pkg/propertyprovider/azure/cloudconfig/config.go [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+		if cfg.UserAssignedIdentityID != "" {
+			return fmt.Errorf("useManagedIdentityExtension needs to be true when userAssignedIdentityID is provided")
+		}
+		if cfg.AADClientID == "" || cfg.AADClientSecret == "" {
+			return fmt.Errorf("AAD client ID or AAD client secret is empty")
+		}
+	}
+
+	// default values
+	if cfg.UserAgent == "" {
```
Comment (jwtty) [inline pkg/propertyprovider/azure/cloudconfig/config.go:113]: no default userAgent

File: pkg/propertyprovider/azure/cloudconfig/test/azure_valid_config.yaml [hunk 13]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (jwtty) [inline pkg/propertyprovider/azure/cloudconfig/test/azure_valid_config.yaml:1]: As per Zhiying's review, we can also add one for an invalid config.

File: pkg/propertyprovider/azure/cloudconfig/config_test.go [hunk 14]
```diff
@@ -0,0 +1,348 @@
...
+				ClusterResourceGroup: "test-rg",
+				CloudProviderBackoff: false,
+			},
+		},
+	}
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			config, err := LoadCloudConfigFromFile(test.filePath)
+			if got := err != nil; got != test.wantErr {
+				t.Fatalf("LoadCloudConfigFromFile() got %v, wantErr %v", err, test.wantErr)
```
Comment (jwtty) [inline pkg/propertyprovider/azure/cloudconfig/config_test.go:341]: `wantErr` is just a bool

File: pkg/propertyprovider/azure/provider.go (hunk 4 above)
Comment (britaniar) [reply pkg/propertyprovider/azure/provider.go:75]: Would that mean we would initialize all the clients we needed too? I also need armnetwork.NewVirtualNetworksClient

File: test/e2e/setup.sh (hunk 2 above)
Comment (zhiying-lin) [reply test/e2e/setup.sh:28]: my bad, misunderstood this variable. thank you for the clarification.

File: pkg/propertyprovider/azure/cloudconfig/config.go [hunk 15]
```diff
@@ -0,0 +1,150 @@
+package cloudconfig
```
Comment (jwtty) [inline pkg/propertyprovider/azure/cloudconfig/config.go:1]: Split this cloud config change in a separate PR. Add a comment before this package. Move this package to /pkg/utils/cloudconfig/azure.

Comment (britaniar) [reply pkg/propertyprovider/azure/cloudconfig/config.go]: Separate PR: https://github.com/Azure/fleet/pull/943

File: charts/member-agent/README.md [hunk 16]
```diff
@@ -29,18 +29,65 @@ helm upgrade member-agent member-agent/ --namespace fleet-system
...
+| propertyProvider   | The property provider to use with the member agent; if none is specified, the Fleet member agent will start with no property provider (i.e., the agent will expose no cluster properties, and collect only limited resource usage information) | ``                                              |
+| region             | The region where the member cluster resides                                                                                                                                                                                                    | ``                                              |
+| config.cloudConfig | The cloud provider configuration                                                                                                                                                                                                               | **required if property provider is enabled**    |
```
Comment (zhiying-lin) [inline charts/member-agent/README.md:45]:
```suggestion
| config.cloudConfig | The cloud provider configuration                                                                                                                                                                                                               | **required if property provider is set to azure**    |
```

File: charts/member-agent/README.md [hunk 17]
```diff
@@ -29,18 +29,65 @@ helm upgrade member-agent member-agent/ --namespace fleet-system
...
+    tenantId: "00000000-0000-0000-0000-000000000000"
+    subscriptionId: "00000000-0000-0000-0000-000000000000"
+    useManagedIdentityExtension: false
+    userAssignedIdentityID: "00000000-0000-0000-0000-000000000000"
+    aadClientId: "00000000-0000-0000-0000-000000000000"
+    aadClientSecret: "<your secret>"
+    userAgent: "fleet-member-agent"
+    resourceGroup: "<resource group name>"
+    location: "<resource group location>"
+    clusterName: "<cluster name>"
```
Comment (jwtty) [inline charts/member-agent/README.md:87]: remove this

File: charts/member-agent/README.md [hunk 18]
```diff
@@ -29,18 +29,65 @@ helm upgrade member-agent member-agent/ --namespace fleet-system
...
+    clusterResourceGroup: "<cluster resource group>"
```
Comment (jwtty) [inline charts/member-agent/README.md:88]: remove this

File: charts/member-agent/README.md [hunk 19]
```diff
@@ -29,18 +29,65 @@ helm upgrade member-agent member-agent/ --namespace fleet-system
...
+| `aadClientId`                 | The ClientID for an AAD application with RBAC access to resources. | Required if `useManagedIdentityExtension` is set to `false`.              |
+| `aadClientSecret`             | The ClientSecret for an AAD application with RBAC access to resources. | Required if `useManagedIdentityExtension` is set to `false`.              |
+| `resourceGroup`               | The name of the resource group where cluster resources are deployed. |                                                                           |
+| `userAgent`                   | The userAgent provided when accessing resources. |                                                                           |
+| `location`                    | The region where resource group and its resources is deployed. |                                                                           |
+| `clusterName`                 | The name of the cluster where the agent is running. |                                                                           |
```
Comment (jwtty) [inline charts/member-agent/README.md:67]: remove this

File: charts/member-agent/README.md [hunk 20]
```diff
@@ -29,18 +29,65 @@ helm upgrade member-agent member-agent/ --namespace fleet-system
...
+| `clusterResourceGroup`        | The resource group where the cluster is deployed. |                                                                           |
```
Comment (jwtty) [inline charts/member-agent/README.md:68]: remove this

File: test/e2e/azure_valid_config.yaml [hunk 21]
```diff
@@ -0,0 +1,11 @@
+config:
+  cloudConfig:
+    cloud: "AzurePublicCloud"
+    tenantId: "00000000-0000-0000-0000-000000000000"
+    subscriptionId: "00000000-0000-0000-0000-000000000000"
+    useManagedIdentityExtension: true
+    clusterResourceGroup: "test-rg"
+    location: "westus"
+    clusterName: "test-cluster"
```
Comment (jwtty) [inline test/e2e/azure_valid_config.yaml:9]: remove this

File: test/e2e/azure_valid_config.yaml [hunk 22]
```diff
@@ -0,0 +1,11 @@
+config:
+  cloudConfig:
+    cloud: "AzurePublicCloud"
+    tenantId: "00000000-0000-0000-0000-000000000000"
+    subscriptionId: "00000000-0000-0000-0000-000000000000"
+    useManagedIdentityExtension: true
+    clusterResourceGroup: "test-rg"
```
Comment (jwtty) [inline test/e2e/azure_valid_config.yaml:7]: remove this

File: charts/member-agent/values.yaml [hunk 23]
```diff
@@ -35,6 +35,21 @@ config:
...
+    useManagedIdentityExtension: false
+    userAssignedIdentityID: ""
+    aadClientId: ""
+    aadClientSecret: ""
+    resourceGroup: ""
+    userAgent: ""
+    location: ""
+    vnetName: ""
+    vnetResourceGroup: ""
+    clusterName: ""
```
Comment (jwtty) [inline charts/member-agent/values.yaml:51]: remove this

File: charts/member-agent/values.yaml [hunk 24]
```diff
@@ -35,6 +35,21 @@ config:
...
+    clusterResourceGroup: ""
```
Comment (jwtty) [inline charts/member-agent/values.yaml:52]: remove this

File: charts/member-agent/templates/deployment.yaml [hunk 25]
```diff
@@ -37,6 +37,9 @@ spec:
             {{- if .Values.propertyProvider }}
             - --property-provider={{ .Values.propertyProvider }}
             {{- end }}
+            {{- if eq .Values.propertyProvider "azure" }}
+            - --cloud-config={{ .Values.cloudConfig }}
```
Comment (jwtty) [inline charts/member-agent/templates/deployment.yaml:41]:
```suggestion
            - --cloud-config=/etc/kubernetes/provider/azure.json
```

File: charts/member-agent/templates/deployment.yaml [hunk 26]
```diff
@@ -84,6 +87,11 @@ spec:
           volumeMounts:
```
Comment (jwtty) [inline charts/member-agent/templates/deployment.yaml:87]: change `{{- if not .Values.useCAAuth }}` to `{{- if or (not .Values.useCAAuth) (eq .Values.propertyProvider "azure") }}` on line 86

File: charts/member-agent/templates/deployment.yaml [hunk 27]
```diff
@@ -84,6 +87,11 @@ spec:
...
           - name: provider-token
```
Comment (jwtty) [inline charts/member-agent/templates/deployment.yaml:88]: Add `{{- if not .Values.useCAAuth }}` and `{{- end }}` round line 88 and 89.

File: charts/member-agent/templates/deployment.yaml [hunk 28]
```diff
@@ -84,6 +87,11 @@ spec:
           volumeMounts:
           - name: provider-token 
             mountPath: /config
+          {{- if eq .Values.propertyProvider "azure" }}
+          - name: cloud-provider-config
+            mountPath: /etc/kubernetes/provider
+            readOnly: true
+          {{- end }}
```
Comment (jwtty) [inline charts/member-agent/templates/deployment.yaml:94]: Add a `{{- end }}` after line 94 to match the `if` on line 86.

File: charts/member-agent/templates/deployment.yaml [hunk 29]
```diff
@@ -105,6 +113,11 @@ spec:
       volumes:
       - name: provider-token
         emptyDir: {}
+        {{- if eq .Values.propertyProvider "azure" }}
```
Comment (jwtty) [inline charts/member-agent/templates/deployment.yaml:116]:
```suggestion
      {{- if eq .Values.propertyProvider "azure" }}
```

File: charts/member-agent/templates/deployment.yaml [hunk 30]
```diff
@@ -105,6 +113,11 @@ spec:
...
+      - name: cloud-provider-config
+        secret:
+          secretName: cloud-config
+        {{- end }}
```
Comment (jwtty) [inline charts/member-agent/templates/deployment.yaml:120]:
```suggestion
      {{- end }}
```

File: charts/member-agent/templates/deployment.yaml [hunk 31]
```diff
@@ -84,6 +87,11 @@ spec:
           volumeMounts:
           - name: provider-token 
             mountPath: /config
+          {{- if eq .Values.propertyProvider "azure" }}
+          - name: cloud-provider-config
+            mountPath: /etc/kubernetes/provider
+            readOnly: true
+          {{- end }}
         - name: refresh-token
```
Comment (jwtty) [inline charts/member-agent/templates/deployment.yaml:95]: Around this container with `{{- if not .Values.useCAAuth }}` and `{{- end }}` after line 112.

File: charts/member-agent/templates/deployment.yaml [hunk 32]
```diff
@@ -105,6 +113,11 @@ spec:
       volumes:
```
Comment (jwtty) [inline charts/member-agent/templates/deployment.yaml:113]: add `{{- if or (not .Values.useCAAuth) (eq .Values.propertyProvider "azure") }}` before this line

File: charts/member-agent/templates/deployment.yaml [hunk 33]
```diff
@@ -105,6 +113,11 @@ spec:
...
       - name: provider-token
```
Comment (jwtty) [inline charts/member-agent/templates/deployment.yaml:114]: around this with `{{- if not .Values.useCAAuth }}` and `{{- end }}` after line 115

File: test/e2e/setup.sh [hunk 34]
```diff
@@ -185,7 +186,9 @@ do
             --set enableV1Alpha1APIs=false \
             --set enableV1Beta1APIs=true \
             --set propertyProvider=$PROPERTY_PROVIDER \
-            --set region=${REGIONS[$i]}
+            --set cloudConfig=$CLOUD_CONFIG_FILE \
```
Comment (jwtty) [inline test/e2e/setup.sh:189]: You do not need this.

File: test/e2e/setup.sh [hunk 35]
```diff
@@ -200,7 +203,9 @@ do
             --set namespace=fleet-system \
             --set enableV1Alpha1APIs=false \
             --set enableV1Beta1APIs=true \
-            --set propertyProvider=$PROPERTY_PROVIDER
+            --set propertyProvider=$PROPERTY_PROVIDER \
+            --set cloudConfig=$CLOUD_CONFIG_FILE \
```
Comment (jwtty) [inline test/e2e/setup.sh:207]: And this is not needed either.

File: charts/member-agent/templates/deployment.yaml [hunk 36]
```diff
@@ -80,10 +83,19 @@ spec:
             httpGet:
               path: /readyz
               port: hubhealthz
-      {{- if not .Values.useCAAuth }}
+      {{- if or (not .Values.useCAAuth) (eq .Values.propertyProvider "azure") }}
```
Comment (jwtty) [inline charts/member-agent/templates/deployment.yaml:86]:
```suggestion
        {{- if or (not .Values.useCAAuth) (eq .Values.propertyProvider "azure") }}
```

File: charts/member-agent/templates/deployment.yaml [hunk 37]
```diff
@@ -80,10 +83,19 @@ spec:
...
+          {{- if not .Values.useCAAuth }}
           - name: provider-token 
             mountPath: /config
+          {{- end }}
+          {{- if eq .Values.propertyProvider "azure" }}
+          - name: cloud-provider-config
+            mountPath: /etc/kubernetes/provider
+            readOnly: true
+          {{- end }}
+      {{- end }}
```
Comment (jwtty) [inline charts/member-agent/templates/deployment.yaml:97]:
```suggestion
        {{- end }}
```

File: test/e2e/setup.sh (hunk 2 above)
Comment (jwtty) [inline test/e2e/setup.sh:28]: This file path is set fixed in our helm chart. We may not need this configuration at all.

**Comment by jwtty** [comment]
LGTM, thanks :)

File: test/e2e/setup.sh [hunk 38]
```diff
@@ -185,7 +185,8 @@ do
             --set enableV1Alpha1APIs=false \
             --set enableV1Beta1APIs=true \
             --set propertyProvider=$PROPERTY_PROVIDER \
-            --set region=${REGIONS[$i]}
+            --set region=${REGIONS[$i]} \
+            -f azure_valid_config.yaml
```
Comment (zhiying-lin) [inline test/e2e/setup.sh:189]: it depends on the property_provider. only when property_provider is azure, we need to use azure_valid_config.yaml

File: test/e2e/setup.sh [hunk 39]
```diff
@@ -200,7 +201,8 @@ do
             --set namespace=fleet-system \
             --set enableV1Alpha1APIs=false \
             --set enableV1Beta1APIs=true \
-            --set propertyProvider=$PROPERTY_PROVIDER
+            --set propertyProvider=$PROPERTY_PROVIDER \
```
Comment (zhiying-lin) [inline test/e2e/setup.sh:204]: ditto

File: cmd/memberagent/main.go [hunk 40]
```diff
@@ -367,6 +369,13 @@ func Start(ctx context.Context, hubCfg, memberConfig *rest.Config, hubOpts, memb
 			klog.V(2).Info("setting up the Azure property provider")
 			// Note that the property provider, though initialized here, is not started until
 			// the specific instance wins the leader election.
+			cloudConfig, err := azureCloudConfig.NewCloudConfigFromFile(*cloudConfigFile)
```
Comment (zhiying-lin) [inline cmd/memberagent/main.go:372]: could you please remove this or add a TODO here as the cloudConfig is not used?

I just notice that that our e2e tests are using property-provider=azure. Need to think about how to test your feature.

If you want to test the feature using this e2e tests, we figure out a way to setup the azure resource correctly. Otherwise, your init azure client call will fail and break the e2e tests.

**Review by zhiying-lin — APPROVED** [review]
